Advanced datetime extraction

qnkhuat opened the pull request:
Taking over the #19275 PR to implement the `temporal-extract` expression function that takes a datetime and a unit.

`[:temporal-extract field :year] => returns the year part of a field`

Because on FE we want to have a simpler syntax, instead of calling `temporalExtract(column, "year")` we want to use `year(column)` or `month(column)` instead.

So I implemented those as sugar clauses and it'll eventually being expanded to `[:temporal-extract field unit]` by the `desugar` middleware.

[Product docs](https://www.notion.so/metabase/Unlock-advanced-date-time-manipulation-038f5cf80f364d7cb93341784169743c)

Follow up:
- [x] [Adds support for DateTime coercion for MongoDB](https://github.com/metabase/metabase/pull/26339)

jeff303 committed:
Implement advanced date/time/zone manipulation, part 1

WORK IN PROGRESS

Incorporate new functions into MBQL and add tests:
 - get-year
 - get-quarter
 - get-month
 - get-day
 - get-day-of-week
 - get-hour
 - get-minute
 - get-second

jeff303 committed:
Fix BigQuery implementations to call extract

Mark as not supported in legacy driver

jeff303 committed:
Merge branch 'master' into advanced-datetime-handling-1-extraction-functions

jeff303 committed:
Add date extraction fns for Postgres

jeff303 committed:
Disable in MongoDB (for now at least)

Disable in BigQuery (legacy driver)

Add implementations for presto-jdbc

jeff303 committed:
Merge branch 'master' into advanced-datetime-handling-1-extraction-functions

qnkhuat committed:
merge with master

qnkhuat committed:
Misc cleanup from Jeff's PR

qnkhuat committed:
Update Jeff's implementation of bigquery-cloud-sqk

qnkhuat committed:
Reorganized tests

qnkhuat committed:
Mongo

qnkhuat committed:
Oracle

qnkhuat committed:
Sqlserver

qnkhuat committed:
Sqlite

qnkhuat committed:
Add casting supports for presto

qnkhuat committed:
Remove Jeff's implementation of presto-jdbc because  its parent is
sql-jdbc

qnkhuat committed:
Update presto-jdbc tests to use the same catalog for all datasets

qnkhuat commented on modules/drivers/presto-jdbc/test/metabase/test/data/presto_jdbc.clj:
```diff
@@ -24,6 +24,12 @@
 ;; during unit tests don't treat presto as having FK support
```
I was getting `Catalog does not exist: database-name` when I tried to test presto-jdbc with my custom dataset, so I updated this to make all dataset for presto-jdbc to use the same catalog.

This is what we did for `presto` too.

qnkhuat committed:
remove unused bindings, requires

qnkhuat committed:
we should be able to use date extract functions in summarization

dpsutton commented:
lint failures:
> /work/test/metabase/query_processor_test/date_time_zone_functions_test.clj:136:24: warning: unused binding driver
/work/test/metabase/query_processor_test/date_time_zone_functions_test.clj:136:31: warning: unused binding col-type

qnkhuat committed:
fix linting

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat committed:
test smarter, not longer

qnkhuat commented on modules/drivers/presto-jdbc/test/metabase/test/data/presto_jdbc.clj:
```diff
@@ -24,6 +24,12 @@
+;; in the past, we had to manually update our Docker image and add a new catalog for every new dataset definition we
+;; added. That's insane. Just use the `test-data` catalog and put everything in that, and use
+;; `db-qualified-table-name` like everyone else.
+(def ^:private test-catalog-name "test_data")
```
I was getting Catalog does not exist: database-name when I tried to test presto-jdbc with my custom dataset, so I updated this to make all dataset for presto-jdbc to use the same catalog.

This is what we did for presto too.

ranquild committed:
Add date extraction functions to the expression editor (#25382)

ranquild commented:
Frontend changes were approved here https://github.com/metabase/metabase/pull/25382

qnkhuat committed:
make sure the semantic type of aggregated columns are integer

qnkhuat committed:
requires mbql.s

qnkhuat committed:
nest doseqs

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

dpsutton commented on src/metabase/driver.clj:
```diff
@@ -462,6 +466,7 @@
 
 (defmethod supports? [::driver :basic-aggregations] [_ _] true)
 (defmethod supports? [::driver :case-sensitivity-string-filter-options] [_ _] true)
+(defmethod supports? [::driver :date-functions] [_ _] true)
```
I'm a little worried that `:date-functions` will be a bit vague for our purposes. We're adding a lot of date functions and we won't be able to make all databases support them. (things like `datediff` and timezone stuff) I wonder if we'll regret the lack of resolution that this imposes on us.

qnkhuat replied:
so `get-xyz` and `date-add, date-subtract` can work on all DBs so I think this is fine. but for `convertTimezone` and `dateDiff` it's might be a different story.

What do you say I rename this to `date-extract` and use this specifically for `get-xyz` functions?

dpsutton replied:
that sounds like a good option. Let's go with that.

dpsutton commented on test/metabase/test/data/interface.clj:
```diff
@@ -163,6 +163,15 @@
+  ;;   Caused by: java.io.FileNotFoundException: Could not locate metabase/driver/all__init.class,
+  ;;   metabase/driver/all.clj or metabase/driver/all.cljc on classpath.
+  ;; add a more explicit check to make it clear what the problem is
+  (when (= :all driver)
```
I'm confused by this addition. I get the idea of it, but was it a problem? If people start with `DRIVERS=foo` they would have the same problem. Is there a reason that `DRIVERS=all` is an expected allowed option?

qnkhuat replied:
ha, this is Jeff's thing. I thought we have a way to set a magic keyword or sth but didn't really look into it. 

Did a quick grep and seems like we don't, so imma removes this change.

dpsutton reviewed (approved):
I'm approving. Had a small note about the enabled check. I'm also curious why this went `get-year`, `get-month` instead of `extract <datetime-expression> :year` so we only have a single method to implement with a few cases in it. Perhaps the autocomplete on the FE can't handle completions of acceptable args? But it always struck me as a bit verbose. Not requesting to change just wondering.

qnkhuat commented:
@dpsutton Yea I think it's for UX purposes since seeing `year([created_at])` is shorter and expression than `extract([created_at], "year")` IMO.

And yes, we currently don't support autocomplete for args in the expression editor.
I'd love to explore this once all functions are implemented because this will become very valuable for users.
Especially for the `convertTimezone(column, to-zone-id, from-zone-id)` function, manually typing in zoneid can be boring and error-prone.

qnkhuat committed:
reverse Jeff's change to test/data/interface

qnkhuat committed:
nested expression and filter using date extraction fns

qnkhuat committed:
rename

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -474,6 +484,9 @@
    (s/recursive #'ArithmeticExpression)
 
+   (partial is-clause? date-extract-functions)
+   (s/recursive #'DatetimeExpression)
```
this will allow us to do `[:- [:get-year field] 3` or `[:abs [:get-year field]]`

qnkhuat committed:
no recursive call in annotate for date-extract func

camsaul reviewed (dismissed):
Having a separate function for each unit is kooky. Nothing else is done that way at all. Let's just have a single `datetime-extract` function that takes unit as a parameter. (The name is just a suggestion. Maybe something else like `temporal-extract` is better... just try to make it clear that it works with dates, times, and datetimes.)

Once that change is in ping me again and I'll review the PR.

qnkhuat committed:
get-unit -> datetime-extract(column, unit)

qnkhuat committed:
desguar nested datetime extraction too

qnkhuat committed:
update schema

qnkhuat committed:
shouldn't be able to create it in the aggregation

qnkhuat committed:
test fixes 🤨

qnkhuat committed:
miss a )

qnkhuat commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -371,18 +371,20 @@
+(defmethod sql.qp/date [:bigquery-cloud-sdk :quarter]          [_ _ expr] (trunc   :quarter   expr))
+(defmethod sql.qp/date [:bigquery-cloud-sdk :quarter-of-year]  [_ _ expr] (extract :quarter   expr))
+(defmethod sql.qp/date [:bigquery-cloud-sdk :year]             [_ _ expr] (trunc   :year      expr))
+(defmethod sql.qp/date [:bigquery-cloud-sdk :yyear]            [_ _ expr] (extract :year      expr))
```
I'm using `yyear` here but I'm open to suggestions. Not sure what `of` a year should be?

dpsutton replied:
We should change this because I don't know what it is supposed to mean. These keywords are internal data so terseness doesn't buy us any benefit.

qnkhuat replied:
yes it should be `temporal-extract column unit` when it reaches the driver.

the main implementation is actually in https://github.com/metabase/metabase/blob/e3ec802b46da320d81219318b9fa744eb8a8c31f/src/metabase/driver/sql/query_processor.clj#L589

becausel call `sql.qp/date` -- a method that dispatch by driver, so everything will automatically works for all sql drivers as long as they implement `sql.qp/date`  function correctly.

dpsutton replied:
ah i follow now.

qnkhuat replied:
I changed `yyear` to `year-of-ear` as @metamben suggested.

qnkhuat committed:
reverse some uncessary chagnes

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat committed:
simplify tests

qnkhuat committed:
datetime-extract -> temporal-extract

qnkhuat commented:
@camsaul @dpsutton I implemented this to have a single function `temporal-extract`.
Because FE is not able to convert from `year` -> `temporal-extract column :year` by itself so I added `get-year, get-month ...` as our sugar clauses and it'll be expanded in the `desugar` middleware.

qnkhuat committed:
fix failed to run presto kerberos

qnkhuat commented on .github/scripts/run-presto-kerberos-integration-test.sh:
```diff
@@ -3,6 +3,7 @@
 set -eo pipefail
 
 # Need Java commands on $PATH, which apparently is not yet the case
+export JAVA_HOME="/usr/lib/jvm/msopenjdk-current"
```
not sure what changed in our Dockerfile but somehow the JAVA_HOME changed.
This CI run just fine 5 days ago and our dockerfile hasn't been updated for over a month now.

![Screen Shot 2022-09-27 at 16 29 02](https://user-images.githubusercontent.com/25661381/192489732-4cf5ded2-c812-4745-8b24-d1507dc5ee9a.png)

qnkhuat replied:
For the record the default `JAVA_HOME=/usr/local/sdkman/candidates/java/current`

dpsutton replied:
This feels a bit wonky to me. The environment for this is setup in `.github/workflows/presto-kerberos-integration-test.yml`. If possible we should ensure the environment is correctly setup there where we declare the env, especially if its a very specific location for the jdk.

qnkhuat replied:
I tried to look into it but that workflow use `mba` which is a CLI that I just know of.
I asked Rai(who built `mba`) and he said had no idea what happened.

 Can we do this in a follow up?

metamben commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -78,6 +78,12 @@
+  "Valid units to extract from a datetime."
+  (s/named
+    (apply s/enum #{:second :minute :hour :day :day-of-week :week :month :quarter :year})
+    "temporal-extract-units"))
```
I wonder why the var name `DateExtractUnits` is different from the schema name `temporal-extract-units`.

qnkhuat replied:
ah, I missed this when `s/date/temporal` the whole thing. Will change.

metamben commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -462,8 +468,20 @@
+
+(def date+time+timezone-functions
+  "Date, time, and timezone related functions."
+  (set/union temporal-extract-functions))
```
What does this `set/union` call achieve?

dpsutton replied:
I think this anticipates follow-up work that adds to these functions.

metamben replied:
I tend to add these when needed. You could add a comment in order no to give readers a pause.

qnkhuat replied:
noted. Sorry about this :)

metamben commented on shared/test/metabase/mbql/util_test.cljc:
```diff
@@ -362,6 +362,12 @@
+    (doseq [[op unit] mbql.u/temporal-extract-ops->unit]
+      (t/is (= [:temporal-extract [:field 1 nil] unit]
+               (mbql.u/desugar-temporal-extract [op [:field 1 nil]]))))))
+
```
Would it make sense to test that de-sugaring works when these expressions are embedded deeper?

metamben commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -109,12 +121,17 @@
+(defmethod date [:sql :second-of-minute] [_driver _ expr] (hx/second expr))
+(defmethod date [:sql :minute-of-hour]   [_driver _ expr] (hx/minute expr))
+(defmethod date [:sql :hour-of-day]      [_driver _ expr] (hx/hour expr))
+(defmethod date [:sql :week-of-year]     [driver _ expr]
```
I wonder why we don't want to have week-of-year extractor.

qnkhuat replied:
I'll check with the product team. if we do I'll add support for it in a follow up.

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -0,0 +1,143 @@
+
+(deftest extraction-function-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :temporal-extract) :mongo)
```
I suggest adding the comment below about mongo here.

metamben reviewed (approved):
Nice work! 👍

qnkhuat commented on frontend/src/metabase/lib/expressions/helper-text-strings.ts:
```diff
@@ -571,6 +571,102 @@ const helperTextStrings: HelpText[] = [
+  {
+    name: "get-day-of-week",
+    structure: "weekday(" + t`column` + ")",
+    description: t`Takes a datetime and returns an integer (1-7) with the number of the day of the week.`,
```
we don't :( this depends on the `first-day-of-week` settings I believe.

qnkhuat committed:
:yyear -> :year-of-era and resolve Tamas's comments

dpsutton reviewed (approved):
tests look a lot better. Fundamentally seems good
